packages/hierarchies-react/src/presentation-hierarchies-react/UseSelectionHandler.ts

```diff
@@ -0,0 +1,165 @@
+  /** Toggles items; allows the use of Ctrl & Shift Keys. */
+  Extended = SelectionModeFlags.KeysEnabled | SelectionModeFlags.DeselectEnabled,
+  /** Disables selection */
+  None = SelectionModeFlags.None,
+}
+
+/** @internal */
+export type SelectionChangeType = "add" | "replace" | "remove";
+
+interface UseSelectionHandlerProps {
+  rootNodes: Array<PresentationTreeNode> | undefined;
+  selectNode: (nodeIds: Array<string>, changeType: SelectionChangeType) => void;
+  selectionMode: SelectionMode;
+}
+
+interface useSelectionHandlerResult {
+  onNodeClick: (nodeId: string, isSelected: boolean, event: React.MouseEvent<HTMLElement, MouseEvent>) => void;
+  onNodeKeyDown: (nodeId: string, isSelected: boolean, event: React.KeyboardEvent<HTMLElement>) => void;
+}
+
+interface FlatTreeState {
+  flatNodeList: Array<string>;
+  nodeIdToIndexMap: Map<string, number>;
+}
+
+/** @beta */
+export function useSelectionHandler(props: UseSelectionHandlerProps): useSelectionHandlerResult {
+  const { rootNodes, selectionMode, selectNode } = props;
+  const selectionModeRef = useLatest(selectionMode);
+  const selectNodeRef = useLatest(selectNode);
+  const previousSelectionRef = useRef<string | undefined>(undefined);
+  const stateRef = useRef<FlatTreeState>({
+    flatNodeList: [],
+    nodeIdToIndexMap: new Map(),
+  });
+
+  useEffect(() => {
+    if (!rootNodes) {
+      return;
+    }
+    stateRef.current = computeFlatNodeList(rootNodes);
+  }, [rootNodes]);
+
+  const getNodeRange = (firstId?: string, secondId?: string) => {
+    const getIndex = (nodeId?: string) => {
+      return nodeId ? stateRef.current.nodeIdToIndexMap.get(nodeId) : 0;
+    };
+    const firstIndex = getIndex(firstId);
+    const secondIndex = getIndex(secondId);
+    if (firstIndex === undefined || secondIndex === undefined) {
+      return [];
+    }
+
+    const startingIndex = Math.min(firstIndex, secondIndex);
+    const endIndex = Math.max(firstIndex, secondIndex);
+    return stateRef.current.flatNodeList.slice(startingIndex, endIndex + 1);
+  };
+
+  const onNodeSelect = (nodeId: string, isSelected: boolean, shiftDown: boolean, ctrlDown: boolean) => {
+    const selection = getSelection(selectionModeRef.current, isSelected, shiftDown, ctrlDown);
+    if (selection.type === "disabled") {
+      return;
+    }
+
+    const nodes = selection.select === "range" ? getNodeRange(previousSelectionRef.current, nodeId) : [nodeId];
+    selection.select !== "range" && (previousSelectionRef.current = nodeId);
+    selectNodeRef.current(nodes, selection.type);
+  };
+
+  const onNodeClick = useRef((nodeId: string, isSelected: boolean, event: React.MouseEvent<HTMLElement, MouseEvent>) => {
+    return onNodeSelect(nodeId, isSelected, event.shiftKey, event.ctrlKey);
+  }).current;
+
+  const onNodeKeyDown = useRef((nodeId: string, isSelected: boolean, event: React.KeyboardEvent<HTMLElement>) => {
+    if (event.key === " " || event.key === "Spacebar" || event.key === "Enter") {
```

Comment:
Do we need `" "` and `"Spacebar"` or one of these should be enough?

Reply:
Based on documentation, older browsers can return "Spacebar" instead of " ".